turbinia/output_writers.py

```diff
@@ -0,0 +1,170 @@
+# Copyright 2015 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Classes to write output to various location types."""
+
+import errno
+import logging
+import os
+import re
+import shutil
+import time
+
+from turbinia import config
+from turbinia import TurbiniaException
+
+from google.cloud import storage
+
+log = logging.getLogger('turbinia')
+
+
+def GetOutputWriters(result):
+  """Get a list of output writers.
+
+  Args:
+    result: A TurbiniaTaskResult object
+
+  Returns:
+    A list of OutputWriter objects.
+  """
+  epoch = str(int(time.time()))
+  log.info(u'%s %s %s' % (epoch, str(result.task_id), result.task_name))
+  unique_dir = u'{0:s}-{1:s}-{2:s}'.format(
+      epoch, str(result.task_id), result.task_name)
+
+  writers = [LocalOutputWriter(base_output_dir=result.base_output_dir,
+                               unique_dir=unique_dir)]
+  config.LoadConfig()
+  if config.GCS_OUTPUT_PATH:
+    writer = GCSOutputWriter(
+        unique_dir=unique_dir, gcs_path=config.GCS_OUTPUT_PATH)
+    writers.append(writer)
+  return writers
+
+
+class OutputWriter(object):
+  """Base class.
+
+  By default this will write the files the Evidence objects point to along with
+  any other files expclicitly written with write().
+
+  Attributes:
+    base_output_dir (string): The base path for output
+    name (string): Name of this output writer
+    unique_dir (string): A psuedo-unique string to be used in paths.
+  """
+
+  def __init__(self, unique_dir=None):
+    """Initialization for OutputWriter."""
+    self.unique_dir = unique_dir
+    self.create_output_dir()
+
+  def create_output_dir(self):
+    """Creates a unique output path for this task and creates directories.
+
+    Needs to be run at runtime so that the task creates the directory locally.
+
+    Returns:
+      A local output path string.
+
+    Raises:
+      TurbiniaException: If there are failures creating the directory.
+    """
+    raise NotImplementedError
+
+  def write(self, file_):
+    """Writes output file.
+
+    Args:
+      file_: A string path to a file.
+
+    Returns:
+      Bool indicating success
+    """
+    raise NotImplementedError
+
+
+class LocalOutputWriter(OutputWriter):
```

Comment:
Not super important at this time but this file will get rather large, maybe we can refactor the different writers out to separate files later?